libs/autotagging-commons/src/main/java/org/opensearch/rule/service/IndexStoredRulePersistenceService.java

```diff
@@ -0,0 +1,266 @@
+                    if (duplicateRuleId != null) {
+                        listener.onFailure(
+                            new IllegalArgumentException(
+                                "A rule that has the same attribute values already exists under rule id " + duplicateRuleId
+                            )
+                        );
+                        return;
+                    }
+                    persistRule(rule, listener);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+        }
+    }
+
+    /**
+     * Persist the rule in the index
+     * @param rule - The rule to update.
+     * @param listener - ActionListener for CreateRuleResponse
+     */
+    public void persistRule(Rule rule, ActionListener<CreateRuleResponse> listener) {
+        try (ThreadContext.StoredContext ctx = getContext()) {
+            IndexRequest indexRequest = new IndexRequest(indexName).source(
+                rule.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS)
+            );
+            client.index(indexRequest, ActionListener.wrap(indexResponse -> {
+                listener.onResponse(new CreateRuleResponse(indexResponse.getId(), rule, RestStatus.OK));
+            }, e -> {
+                logger.warn("Failed to save Rule object due to error: {}", e.getMessage());
+                listener.onFailure(e);
+            }));
+        } catch (IOException e) {
+            logger.error("Error saving rule to index: {}", indexName, e);
+            listener.onFailure(new RuntimeException("Failed to save rule to index."));
+        }
+    }
+
+    /**
+     * Creates the system index .rules if it doesn't exist
+     * @param listener - ActionListener for CreateRuleResponse
+     */
+    void createIndexIfAbsent(ActionListener<Boolean> listener) {
+        try (ThreadContext.StoredContext ctx = getContext()) {
+            if (clusterService.state().metadata().hasIndex(indexName)) {
+                listener.onResponse(true);
+                return;
+            }
+            final CreateIndexRequest createIndexRequest = new CreateIndexRequest(indexName).settings(indexSettings);
+            client.admin().indices().create(createIndexRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(CreateIndexResponse response) {
+                    logger.info("Index {} created?: {}", indexName, response.isAcknowledged());
+                    listener.onResponse(response.isAcknowledged());
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof ResourceAlreadyExistsException) {
+                        logger.info("Index {} already exists", indexName);
```

Comment:
I think the logging info should be set to trace for such repetitive log statements